ui/src/components/WalletConnection.jsx

```diff
@@ -0,0 +1,55 @@
+import { makeReactAgoricWalletConnection } from '@agoric/wallet-connection/react.js';
+import React, { useCallback } from 'react';
+import { E } from '@agoric/eventual-send';
+
+import { dappConfig } from '../utils/config';
+import { setApproved, setConnected } from '../store';
+
+// Create a wrapper for agoric-wallet-connection that is specific to
+// the app's instance of React.
+const AgoricWalletConnection = makeReactAgoricWalletConnection(React);
+
+const WalletConnection = ({ setWalletP, dispatch }) => {
+  const { CONTRACT_NAME } = dappConfig;
+
+  const onWalletState = useCallback(ev => {
+    const { walletConnection, state } = ev.detail;
+    switch (state) {
+      case 'idle': {
+        // This is one of the only methods that the wallet connection facet allows.
+        // It connects asynchronously, but you can use promise pipelining immediately.
+        /** @type {ERef<WalletBridge>} */
+        const bridge = E(walletConnection).getScopedBridge(CONTRACT_NAME);
+        // You should reconstruct all state here.
+        setWalletP(bridge);
+        break;
+      }
+      case 'approving': {
+        dispatch(setApproved(false));
+        break;
+      }
+      case 'bridged': {
+        dispatch(setConnected(true));
+        dispatch(setApproved(true));
```

Comment:
It's a bit tricky because the state goes like:
idle -> bridged -> approving -> bridged

Maybe we should add a state for when it's bridged but the approval status is unknown? Otherwise it's a bit tricky to rely on this to know whether the dapp needs approval or not, there's some back-and-forth between states.

Reply:
> Maybe we should add a state for when it's bridged but the approval status is unknown? 

I agree.  The issue is that the approval may be revoked at any time, and we need to go back to `approving` in that case.   But definitely should not go to `bridged` first before `approving`.  That's probably fairly simple to fix.